src/Riok.Mapperly/Descriptors/Mappings/DerivedExistingTargetTypeMapping.cs

```diff
@@ -0,0 +1,83 @@
+using Riok.Mapperly.Emit.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using static Riok.Mapperly.Emit.Syntax.SyntaxFactoryHelper;
+
+namespace Riok.Mapperly.Descriptors.Mappings;
+
+/// <summary>
+/// A derived type mapping maps one base type or interface to another
+/// by implementing a if with instance checks over known types and performs the provided mapping for each type.
+/// </summary>
+public class DerivedExistingTargetTypeMapping : ExistingTargetMapping
+{
+    private const string SourceName = "source";
+    private const string TargetName = "target";
+    private const string GetTypeMethodName = nameof(GetType);
+    private readonly IReadOnlyCollection<IExistingTargetMapping> _existingTargetTypeMappings;
+
+    public DerivedExistingTargetTypeMapping(
+        ITypeSymbol sourceType,
+        ITypeSymbol targetType,
+        IReadOnlyCollection<IExistingTargetMapping> existingTargetTypeMappings
+    )
+        : base(sourceType, targetType)
+    {
+        _existingTargetTypeMappings = existingTargetTypeMappings;
+    }
+
+    public override IEnumerable<StatementSyntax> Build(TypeMappingBuildContext ctx, ExpressionSyntax target)
+    {
+        var sourceExpression = TupleExpression(CommaSeparatedList(Argument(ctx.Source), Argument(target)));
+        var sections = _existingTargetTypeMappings.Select(x => BuildSwitchSection(ctx, x)).Append(BuildDefaultSwitchSection(ctx, target));
+
+        yield return ctx.SyntaxFactory.SwitchStatement(sourceExpression, List(sections)).AddLeadingLineFeed(ctx.SyntaxFactory.Indentation);
+    }
+
+    private SwitchSectionSyntax BuildSwitchSection(TypeMappingBuildContext ctx, IExistingTargetMapping mapping)
+    {
+        var (sectionCtx, sourceVariableName) = ctx.WithNewScopedSource(SourceName);
+        var targetVariableName = sectionCtx.NameBuilder.New(TargetName);
+        sectionCtx = sectionCtx.AddIndentation();
+
+        // (A source, B target)
+        var positionalTypeMatch = PositionalPatternClause(
+            CommaSeparatedList(
+                Subpattern(DeclarationPattern(mapping.SourceType, sourceVariableName)),
+                Subpattern(DeclarationPattern(mapping.TargetType, targetVariableName))
+            )
+        );
+        var pattern = RecursivePattern().WithPositionalPatternClause(positionalTypeMatch);
+
+        // case (A source, B target):
+        var caseLabel = CasePatternSwitchLabel(pattern).AddLeadingLineFeed(sectionCtx.SyntaxFactory.Indentation);
+
+        // break;
+        var statementContext = sectionCtx.AddIndentation();
+        var breakStatement = BreakStatement().AddLeadingLineFeed(statementContext.SyntaxFactory.Indentation);
+        var target = IdentifierName(targetVariableName);
+        var statements = List(mapping.Build(statementContext, target).Append(breakStatement));
+
+        return SwitchSection().WithLabels(SingletonList<SwitchLabelSyntax>(caseLabel)).WithStatements(statements);
```

Comment:
I'd move this to the syntax factory. This could then look like this:
```c#
var caseLabel = ...;
var statements = ...;
return SwitchSection(caseLabel, statements);
```

The switch section syntax helper constructs the `SwitchSection` and adds the break statement. An additional `SwitchDefaultSection(statements)` could also be helpful.